### ISSUE TITLE
No Support v4 SwipeToRefreshLayout Shadow

### ISSUE THREAD

**Issue by tjohn** [robolectric/robolectric#1830]
Hey so I was going to implement a Shadow class for SwipeToRefreshLayout, is this something that the community wants. The behavior that I want to test is actually triggering the refresh listener which cannot be accessed via the standard API. I will be building this out for my project and just wanted to check if this would be useful for others before going through with a PR.

**Comment by jongerrish** [comment]
This sounds like a useful shadow, I say go for it!

**Comment by jaredsburrows** [comment]
Why are you making a `shadow` for this? You just want to test the `SwipeRefreshLayout.OnRefreshListener`?

**Comment by tjohn** [comment]
I want to test that in my activity/fragment when the swipe happens an API request is made. A little integrationy, but I think this is a valid test case, just like clicking on a view is valid to test the `onClickListener`.

**Comment by erd** [comment]
I'm sure someone would find it useful. Submit a PR and I'll take a look at it.

**Comment by jaredsburrows** [comment]
@tjohn  Check out: https://github.com/hidroh/materialistic/blob/master/app/src/test/java/io/github/hidroh/materialistic/test/ShadowSwipeRefreshLayout.java

**Comment by tjohn** [comment]
that would work:

```java
@Implements(SwipeRefreshLayout.class)
public class ShadowSwipeRefreshLayout extends ShadowViewGroup {
  @RealObject SwipeRefreshLayout realObject;

  private SwipeRefreshLayout.OnRefreshListener listener;

  @Implementation
  public void setOnRefreshListener(SwipeRefreshLayout.OnRefreshListener listener) {
    this.listener = listener;
    Shadow.directlyOn(realObject, SwipeRefreshLayout.class).setOnRefreshListener(listener);
  }

  public void refresh() {
    listener.onRefresh();
  }
}
```

is what I did. @erd do you have a preference?

**Comment by jaredsburrows** [comment]
@tjohn Are you going to make this PR?

**Comment by tjohn** [comment]
It was unclear which PR was preferred. I can totally submit mine, but
wasn't sure if the project prefers the other option.

Trevor

**Comment by erd** [comment]
> I can totally submit mine, but wasn't sure if the project prefers the other option.

I like the fact that yours saves the listener and calls into the underlying method. The one change I would make is to have a `getOnRefreshListener` instead of `refresh`. Just to keep it simple.

**Comment by tjohn** [comment]
OK I'll submit it today.

Trevor

**Comment by jaredsburrows** [comment]
+1

**Comment by tjohn** [comment]
Ah so I am having an issue updating the support library. Not a great maven user and when I try to update it I am just in dependency hell. I have tried updating the install-dependencies script and the pom, but can't get intellij to sort it out. If anyone has any tips that would be helpful. It also seems like updating the support library should be a separate PR.

**Comment by jaredsburrows** [comment]
@tjohn @erd This is why we need to get this to work on `Gradle` and remove the `Maven` branch.

**Comment by erd** [comment]
@jaredsburrows You're welcome to work on it. It's not a priority for me.

**Comment by jaredsburrows** [comment]
@erd  I would like to. Do you care if the directory structure changed some?

For instance, 4 `Gradle` modules:

- `org.robolectric:robolectric-core:3.0 `                          // <-- all modules rely on the "core"
- `org.robolectric:robolectric-shadows-support-v4:3.0`
- `org.robolectric:robolectric-shadows-httpclient:3.0`
- `org.robolectric:robolectric-shadows-maps:3.0`

**Comment by jaredsburrows** [comment]
@erd  Also, consider merging @lexs PR here, https://github.com/robolectric/robolectric/pull/1840. Before I work on moving to gradle. It is quite large.

**Comment by gkylafas** [comment]
Is this going to be included in a 3.0-rcX release? I am using 3.0-rc3 currently and I just stumbled upon the need for a SwipeToRefreshLayoutShadow.

**Comment by tjohn** [comment]
I cannot finish the PR from my side, so I am guessing no. Right now I am
just using a custom shadow.

Trevor
